klippa/src/serialize.rs

```diff
@@ -0,0 +1,378 @@
+    tail: usize,
+    real_links: Vec<Link>,
+    virtual_links: Vec<Link>,
+    next_obj: usize,
+}
+
+impl Default for Object {
+    fn default() -> Object {
+        Object {
+            head: 0,
+            tail: 0,
+            real_links: vec![],
+            virtual_links: vec![],
+            next_obj: NULL_IDX,
+        }
+    }
+}
+
+type ObjIdx = u32;
+#[allow(dead_code)]
+const INVALID_OBJ_IDX: ObjIdx = ObjIdx::MAX;
+
+#[allow(dead_code)]
+#[derive(Default)]
+struct Link {
+    width: u8,
+    is_signed: bool,
+    whence: u8,
+    bias: u32,
+    position: u32,
+    objidx: ObjIdx,
+}
+
+#[derive(Default)]
+#[allow(dead_code)]
+pub(crate) struct Serializer {
+    pub(crate) start: usize,
+    pub(crate) end: usize,
+    pub(crate) head: usize,
+    pub(crate) tail: usize,
+    // TODO: zerocopy, debug_depth
+    pub(crate) errors: SerializeErrorFlags,
+
+    data: Vec<u8>,
+    object_pool: ObjectPool,
+    // index for current Object in the object_pool
+    current: usize,
+
+    packed: Vec<ObjIdx>,
+    packed_map: FnvHashMap<usize, ObjIdx>,
+}
+
+#[allow(dead_code)]
+impl Serializer {
+    pub(crate) fn new(size: u32) -> Self {
+        let buf_size = size as usize;
+        let mut this = Serializer {
+            data: vec![0; buf_size],
+            end: buf_size,
+            tail: buf_size,
+            current: NULL_IDX,
+            packed: Vec::new(),
+            ..Default::default()
+        };
+
+        this.packed.push(INVALID_OBJ_IDX);
+        this
+    }
+
+    // Embed a single Scalar type
+    pub(crate) fn embed(&mut self, obj: impl Scalar) -> Result<usize, SerializeErrorFlags> {
+        let raw = obj.to_raw();
+        let bytes = raw.as_ref();
+        let size = bytes.len();
+
+        let ret = self.allocate_size(size)?;
+        self.data
+            .get_mut(ret..ret + size)
+            .unwrap()
+            .copy_from_slice(bytes);
+        Ok(ret)
+    }
+
+    // Allocate size
+    pub(crate) fn allocate_size(&mut self, size: usize) -> Result<usize, SerializeErrorFlags> {
+        if self.in_error() {
+            return Err(self.errors);
+        }
+
+        if size > u32::MAX as usize || self.tail - self.head < size {
+            return Err(self.err(SerializeErrorFlags::SERIALIZE_ERROR_OUT_OF_ROOM));
+        }
+
+        //TODO: add support for clear?
+        let ret = self.head;
+        self.head += size;
+        Ok(ret)
+    }
+
+    #[inline]
+    pub(crate) fn successful(&self) -> bool {
+        !self.errors
+    }
+
+    #[inline]
+    pub(crate) fn in_error(&self) -> bool {
+        !!self.errors
+    }
+
+    #[inline]
+    pub(crate) fn err(&mut self, error_type: SerializeErrorFlags) -> SerializeErrorFlags {
+        self.errors |= error_type;
+        self.errors
+    }
+
+    pub(crate) fn push(&mut self) -> Result<(), SerializeErrorFlags> {
+        if self.in_error() {
+            return Err(self.errors);
+        }
+
+        let obj_idx = self.object_pool.alloc();
```

Comment:
The term object index is being used in two places with different meanings:
- Here meaning the index in the object pool
- Second, in Link's to represent the index in the packed array.

To avoid confusion I suggest giving these two things distinct types:

PackedIndex and PoolIndex
or
ObjIndex and PoolIndex if you want to stay closer to hb terminology.